### PR TITLE
Adding Glue Data Quality Rule Recommendation Run 

### PR DESCRIPTION
**Pull request by gopidesupavan**
Adding Amazon Glue Data Quality Recommendation Run. Doc, Operator, Sensor, Trigger, Waiter, Unit Test, System Test.

Manually tested in Breeze with
 1. wait_for_completion=False with a Sensor
 2. deferrable=True. 
 3. wait_for_completion=True
 
Please Don't mind the first one failed due to wrong table i have provided.

image

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/glue/client/start_data_quality_rule_recommendation_run.html

Sample Dag for recommendation run:

```
from datetime import datetime

from airflow import DAG
from airflow.providers.amazon.aws.operators.glue import GlueDataQualityRuleRecommendationRunOperator

with DAG(
    dag_id="example_glue_data_quality",
    schedule="@once",
    start_date=datetime(2021, 1, 1),
    tags=["example_recommendation_test"],
    catchup=False,
) as dag:
    start_recommendation_run = GlueDataQualityRuleRecommendationRunOperator(
        task_id="start_recommendation_run",
        datasource={
            "GlueTable": {
                "TableName": "test_table",
                "DatabaseName": "test_default",
            }
        },
        role="arn:aws:iam::{ACCOUNT_ID}:role/GlueDataQuality",
        recommendation_run_kwargs={"CreatedRulesetName": "test-rule-set"},
    )

```

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/operators/glue.py [hunk 1]
```diff
@@ -499,3 +500,149 @@ def execute_complete(self, context: Context, event: dict[str, Any] | None = None
...
+        try:
+            response = self.hook.conn.start_data_quality_rule_recommendation_run(
+                DataSource=self.datasource,
+                Role=self.role,
+                NumberOfWorkers=self.number_of_workers,
+                Timeout=self.timeout,
+                **self.recommendation_run_kwargs,
+            )
+        except ClientError as error:
+            print(error)
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/glue.py]: Left over? Or if it is intended, we should log instead

File: airflow/providers/amazon/aws/operators/glue.py [hunk 2]
```diff
@@ -499,3 +500,149 @@ def execute_complete(self, context: Context, event: dict[str, Any] | None = None
...
+            )
+            self.log.info(
+                "AWS Glue data quality recommendation run completed RunId: %s", recommendation_run_id
+            )
+
+            if self.show_results:
+                self.hook.log_recommendation_results(run_id=recommendation_run_id)
+
+        else:
+            self.log.info("AWS Glue data quality recommendation run runId: %s.", recommendation_run_id)
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/glue.py]:
```suggestion
            self.log.info(message_description)
```

File: tests/system/providers/amazon/aws/example_glue_data_quality_with_recommendation.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    GlueDataQualityRuleRecommendationRunSensor,
+    GlueDataQualityRuleSetEvaluationRunSensor,
+)
+from airflow.utils.trigger_rule import TriggerRule
+from tests.system.providers.amazon.aws.utils import SystemTestContextBuilder
+
+ROLE_ARN_KEY = "ROLE_ARN"
+sys_test_context_task = SystemTestContextBuilder().add_variable(ROLE_ARN_KEY).build()
+
+DAG_ID = "example_glue_data_quality_recommendation_run"
```
Comment (vincbeck) [inline tests/system/providers/amazon/aws/example_glue_data_quality_with_recommendation.py]: We usually set the DAG ID as the filename

Comment (gopidesupavan) [reply tests/system/providers/amazon/aws/example_glue_data_quality_with_recommendation.py]: Sure, made the changes.

File: airflow/providers/amazon/aws/operators/glue.py (hunk 2 above)
Comment (gopidesupavan) [reply airflow/providers/amazon/aws/operators/glue.py]: Made the changes. thanks

File: airflow/providers/amazon/aws/operators/glue.py (hunk 1 above)
Comment (gopidesupavan) [reply airflow/providers/amazon/aws/operators/glue.py]: Good catch, Sorry my bad i left it there. i have removed now. thank you.

**Comment by gopidesupavan** [comment]
Thank you @vincbeck for review, i have made all the changes.

**Review by vincbeck — APPROVED** [review]

**Comment by syedahsn** [comment]
I'm trying to run the system test locally, before I get have it running in our system test dashboard, but I'm getting some errors. 
In the `create_database` task, I get the following error:
```
[2024-06-10, 22:50:41 UTC] {taskinstance.py:3131} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/opt/airflow/airflow/models/taskinstance.py", line 759, in _execute_task
    result = _execute_callable(context=context, **execute_callable_kwargs)
  File "/opt/airflow/airflow/models/taskinstance.py", line 725, in _execute_callable
    return ExecutionCallableRunner(
  File "/opt/airflow/airflow/utils/operator_helpers.py", line 250, in run
    return self.func(*args, **kwargs)
  File "/opt/airflow/airflow/models/baseoperator.py", line 406, in wrapper
    return func(self, *args, **kwargs)
  File "/opt/airflow/airflow/providers/amazon/aws/operators/athena.py", line 131, in execute
    self.query_execution_id = self.hook.run_query(
  File "/opt/airflow/airflow/providers/amazon/aws/hooks/athena.py", line 133, in run_query
    response = self.get_conn().start_query_execution(**params)
  File "/usr/local/lib/python3.9/site-packages/botocore/client.py", line 565, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python3.9/site-packages/botocore/client.py", line 1021, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.InvalidRequestException: An error occurred (InvalidRequestException) when calling the StartQueryExecution operation: line 1:8: mismatched input 'DATABASE'. Expecting: 'MATERIALIZED', 'MULTI', 'OR', 'PROTECTED', 'ROLE', 'SCHEMA', 'TABLE', 'VIEW'
```
This is for both the system test in this PR (`example_glue_data_quality_with_recommendation.py`), and the system test from #39923 (`example_glue_data_quality.py`). 
@gopidesupavan were you able to run the system test locally to completion?

**Comment by gopidesupavan** [comment]
Hi @syedahsn Yes, have tested locally. Just triggered one run for both the example_glue_data_quality_with_recommendation and example_glue_data_quality the outputs are below.

```
pytest --system amazon tests/system/providers/amazon/aws/example_glue_data_quality_with_recommendation.py --capture=no

[2024-06-11T06:06:39.992+0000] {__init__.py:47} INFO - EXECUTION SUMMARY:
╒═══════════════════════════════════════════════════════════════════════════╤══════════╕
│ Task ID                                                                   │ Status   │
╞═══════════════════════════════════════════════════════════════════════════╪══════════╡
│ variable_fetcher                                                          │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ create_s3_bucket                                                          │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ upload_sample_data                                                        │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ create_database                                                           │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ create_table                                                              │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ glue_data_quality_recommendation_workflow.recommendation_run              │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ glue_data_quality_recommendation_workflow.await_recommendation_run_sensor │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ glue_data_quality_recommendation_workflow.start_evaluation_run            │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ glue_data_quality_recommendation_workflow.await_evaluation_run_sensor     │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ delete_ruleset                                                            │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ drop_table                                                                │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ drop_database                                                             │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ delete_s3_bucket                                                          │ success  │
├───────────────────────────────────────────────────────────────────────────┼──────────┤
│ watcher                                                                   │ skipped  │
╘═══════════════════════════════════════════════════════════════════════════╧══════════╛
[2024-06-11T06:06:39.993+0000] {dagrun.py:900} INFO - DagRun Finished: dag_id=example_glue_data_quality_with_recommendation, execution_date=2024-06-11 06:00:20.851722+00:00, run_id=manual__2024-06-11T06:00:20.851722+00:00, run_start_date=2024-06-11 06:00:20.851722+00:00, run_end_date=2024-06-11 06:06:39.977671+00:00, run_duration=379.125949, state=success, external_trigger=False, run_type=manual, data_interval_start=2024-06-11 06:00:20.851722+00:00, data_interval_end=2024-06-11 06:00:20.851722+00:00, dag_hash=None
PASSED

pytest --system amazon tests/system/providers/amazon/aws/example_glue_data_quality.py --capture=no

[2024-06-11T06:13:16.444+0000] {__init__.py:47} INFO - EXECUTION SUMMARY:
╒════════════════════════════════════════════════════════╤══════════╕
│ Task ID                                                │ Status   │
╞════════════════════════════════════════════════════════╪══════════╡
│ variable_fetcher                                       │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ create_s3_bucket                                       │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ upload_sample_data                                     │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ create_database                                        │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ create_table                                           │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ glue_data_quality_workflow.create_rule_set             │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ glue_data_quality_workflow.start_evaluation_run        │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ glue_data_quality_workflow.await_evaluation_run_sensor │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ delete_ruleset                                         │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ drop_table                                             │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ drop_database                                          │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ delete_s3_bucket                                       │ success  │
├────────────────────────────────────────────────────────┼──────────┤
│ watcher                                                │ skipped  │
╘════════════════════════════════════════════════════════╧══════════╛
[2024-06-11T06:13:16.444+0000] {dagrun.py:900} INFO - DagRun Finished: dag_id=example_glue_data_quality, execution_date=2024-06-11 06:08:57.055240+00:00, run_id=manual__2024-06-11T06:08:57.055240+00:00, run_start_date=2024-06-11 06:08:57.055240+00:00, run_end_date=2024-06-11 06:13:16.428124+00:00, run_duration=259.372884, state=success, external_trigger=False, run_type=manual, data_interval_start=2024-06-11 06:08:57.055240+00:00, data_interval_end=2024-06-11 06:08:57.055240+00:00, dag_hash=None
PASSED


```

**Comment by gopidesupavan** [comment]
@syedahsn  Are you setting env_id manually? I am using env_id for the database name and tablename. if it has anything `-` in the env_id it may cause issue,. I guess Athena wont allow table and database name with char `-` . Please could you check.

**Comment by syedahsn** [comment]
Yep, it turned out that the issue was a `-` in the env id. Everything seems to be working correctly, and we should have these 2 system tests running in our suite of tests shortly. Thanks for looking into it!

**Comment by gopidesupavan** [comment]
Thanks for confirming.

**Pull request merged**